### PR TITLE
Migrates the WCS unit test helper classes/functions to subscriptions core

### LINKED ISSUE #60: Migrating Unit Tests: Unit test helper classes
## Description

The goal of this issue is to move the [subscriptions unit test helper classes/functions](https://github.com/woocommerce/woocommerce-subscriptions/tree/trunk/tests/framework/helpers) into the `subscriptions-core` repository. These include:
 - [x] `class-wcs-helper-coupon.php`
 - [x] `class-wcs-helper-product.php`
 - [x] `class-wcs-helper-subscription.php`
 - [x] `class-wcs-helper-upgrade-repair.php`

These helpers are used across most of our unit tests so they'll need to be moved first.

As part of moving this functionality over, there might be some things we could take/learn from WC Payments helpers (i.e. WC data store mock and WC order helpers etc)

I'm open to ideas on this but I'm thinking we also move these to `tests/unit/helpers` to keep it consistent with WC Payments.

---

Related to https://github.com/woocommerce/woocommerce-subscriptions/issues/4195

### PR DESCRIPTION
**Pull request by mattallan**
Fixes #60

## Description

This PR migrates the `woocommerce-subscriptions/tests/framework/helpers` over to the subscriptions core codebase so that we can use them in our new unit tests.

During the move, I changed a few minor things with these files:
 - Removed WC pre 3.x compatibility code
 - Updated arrays to be written as `[]` instead of `array()`
 - Cleaned up some of the doc blocks and whitespace
 - Changed the path of the helpers to be consistent with WC Payments (`tests/unit/helpers`)
 - `WCS_Helper_Coupon` no longer extends `WC_Helper_Coupon`

## How to test this PR

To test this I just tried using some of our helpers in the `Simple_Pass_Test->test_always_pass()` to check they're working.

### CODE REVIEW

**Comment by mattallan** [comment]
Hey @brucealdridge, I noticed one of the cart functions tests is failing only on PHP7.0  with an error that doesn't make much sense to me 🤔 

Do you have any idea if this issue is:
 1. The unit test doesn't work on PHP7.0
 2. The `wcs_get_cart_item_name()` doesn't work on PHP7.0 (this is more concerning)
 
I had a look at the function but nothing stood out as problematic.

File: composer.json [hunk 1]
```diff
@@ -12,7 +12,10 @@
 		"phpunit/phpunit": "^6.0",
 		"woocommerce/woocommerce-sniffs": "0.1.0",
 		"dave-liddament/sarb": "^1.1",
-		"yoast/phpunit-polyfills": "^1.0"
+		"yoast/phpunit-polyfills": "^1.0",
+		"phpdocumentor/reflection-docblock": "4.3.4",
+		"phpdocumentor/type-resolver": "0.5.1",
+		"doctrine/instantiator": "1.0.5"
```
Comment (shendy-a8c) [inline composer.json:18]: I can't put comment in a JSON file, so putting description here.

These packages (`phpdocumentor/reflection-docblock`, `phpdocumentor/type-resolve`, and `doctrine/instantiator`) are dependencies of `phpspec/prophecy`.
`phpspec/prophecy` is dependency of `phpunit/phpunit`.

We're using `phpunit/phpunit` `6.5.14`. Even though it supports PHP 7.0, turns out, `phpdocumentor/reflection-docblock`, `phpdocumentor/type-resolve`, and `doctrine/instantiator` are in versions that do not support PHP 7.0, causing fatal error when running unit test.

Specifically, the error is about constant (`const`) visibility (`public`, `protected`, or `private`) that is only introduced in PHP 7.1. Requiring those packages here with the version to make sure unit test isn't broken when run with PHP 7.0.

File: tests/unit/helpers/class-wc-helper-order.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+		}
+
+		WC_Helper_Shipping::delete_simple_flat_rate();
+
+		// Delete the order post.
+		$order->delete( true );
+	}
+
+	/**
+	 * Create a order.
```
Comment (shendy-a8c) [inline tests/unit/helpers/class-wc-helper-order.php:36]: nit:
```suggestion
	 * Create an order.
```

File: tests/unit/helpers/class-wcs-helper-subscription.php [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+			$billing_data
+		);
+		$order->set_address( $billing_address, 'billing' );
+
+		// Add shipping costs
+		self::add_shipping( $order );
+
+		// Set payment gateway
+		$payment_gateways = WC()->payment_gateways->payment_gateways();
+		$order->set_payment_method( $payment_gateways['bacs'] );
```
Comment (shendy-a8c) [inline tests/unit/helpers/class-wcs-helper-subscription.php:194]: Just curious, what's `bacs`? Direct transfer?

**Comment by brucealdridge** [comment]
> Hey @brucealdridge, I noticed one of the cart functions tests is failing only on PHP7.0 with an error that doesn't make much sense to me 🤔

Sorry I didn't get to this,  thanks @shendy-a8c for saving the day!!

**Comment by mattallan** [comment]
Thanks @brucealdridge and @shendy-a8c ! I believe all the helpers and bootstrap related work is done so I'm going to merge this one to avoid a huge PR and we can continue reviewing the others separately (#71 and #73)

🙌

**Review by shendy-a8c — APPROVED** [review]
I forgot to submit my review last week!

Overall, it looks good. Just left some comment for nit.

**Pull request merged**